src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+                          DefaultInitVec<size_t> const &mappings) {
+	std::filebuf output;
+	output.open(options.attrmap, std::ios_base::out | std::ios_base::binary);
+
+	assert(options.baseTileIDs[0] == 0 && options.baseTileIDs[1] == 0); // TODO: deal with offset
+	for (AttrmapEntry const &entry : attrmap) {
+		uint8_t attr = entry.xFlip << 5 | entry.yFlip << 6;
+		attr |= (entry.tileID >= options.maxNbTiles[0]) << 3;
+		attr |= mappings[entry.protoPaletteID] & 7;
+		output.sputc(attr);
+	}
+}
+
+} // namespace optimized
+
+void process() {
+	options.verbosePrint("Using libpng %s\n", png_get_libpng_ver(nullptr));
+
+	options.verbosePrint("Reading tiles...\n");
+	Png png(options.input);
+	ImagePalette const &colors = png.getColors();
+
+	// Now, we have all the image's colors in `colors`
+	// The next step is to order the palette
+
+	if (options.beVerbose) {
+		options.verbosePrint("Image colors: [ ");
+		size_t i = 0;
+		for (auto const &slot : colors) {
+			if (!slot.has_value()) {
+				continue;
+			}
+			options.verbosePrint("#%02x%02x%02x%02x%s", slot->red, slot->green, slot->blue,
+			                     slot->alpha, i != colors.size() - 1 ? ", " : "");
+			++i;
+		}
+		options.verbosePrint("]\n");
+	}
+
+	// Now, iterate through the tiles, generating proto-palettes as we go
+	// We do this unconditionally because this performs the image validation (which we want to
+	// perform even if no output is requested), and because it's necessary to generate any
+	// output (with the exception of an un-duplicated tilemap, but that's an acceptable loss.)
+	std::vector<ProtoPalette> protoPalettes;
+	DefaultInitVec<AttrmapEntry> attrmap{};
+
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		ProtoPalette tileColors;
+		AttrmapEntry &attrs = attrmap.emplace_back();
+
+		for (uint32_t y = 0; y < 8; ++y) {
+			for (uint32_t x = 0; x < 8; ++x) {
+				tileColors.add(tile.pixel(x, y));
+			}
+		}
+
+		// Insert the palette, making sure to avoid overlaps
+		// TODO: if inserting (0, 1), (0, 2), and then (0, 1, 2), we might have a problem!
+		for (size_t i = 0; i < protoPalettes.size(); ++i) {
+			switch (tileColors.compare(protoPalettes[i])) {
+			case ProtoPalette::WE_BIGGER:
+				protoPalettes[i] = tileColors; // Override them
+				[[fallthrough]];
+			case ProtoPalette::THEY_BIGGER:
+				// Do nothing, they already contain us
+				attrs.protoPaletteID = i;
+				goto contained;
+			case ProtoPalette::NEITHER:
+				break; // Keep going
+			}
+		}
+		attrs.protoPaletteID = protoPalettes.size();
+		protoPalettes.push_back(tileColors);
+contained:;
+	}
+
+	options.verbosePrint("Image contains %zu proto-palette%s\n", protoPalettes.size(),
+	                     protoPalettes.size() != 1 ? "s" : "");
+
+	// Sort the proto-palettes by size, which improves the packing algorithm's efficiency
+	// TODO: try keeping the palettes stored while inserting them instead, might perform better
```

Comment:
[Probably not](https://stackoverflow.com/a/2710332):

> The reason such a container is not part of the standard library is that it would be inefficient. Using a vector for storage means objects have to be moved if something is inserted in the middle of the vector. Doing this on every insertion gets needlessly expensive. (On average, half the objects will have to be moved for each insertion. That's pretty costly)
> 
> If you want a sorted vector, it is likely better to insert all the elements, and then call `std::sort()` once, after the insertions.

Reply:
Will add a note